### PR TITLE
Add classification head for segmentation model

### PR DESCRIPTION
**Pull request by qlzh727**
# What does this PR do?
Add classification head for segmentation model

This will be the last piece before we ship the deeplab v3 with KCV components only.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [X] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

### CODE REVIEW

File: keras_cv/models/segmentation/__internal__/segmentation_head.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class SegmentationHead(layers.Layer):
+    """Prediction head for the segmentation model
+
+    The head will take the output from decoder (eg FPN or ASPP), and produce production
+    (pixel level classifications) as the output for the model.
+
+    Args:
+        classes: int, the number of output classes for the prediction.
```
Comment (tanzhenyu) [inline keras_cv/models/segmentation/__internal__/segmentation_head.py:27]: Can we help users by clarifying whether this will include background class? It's likely we require 1+num_foreground_classes here, but user pass num_foreground_classes?
(assuming we will make this as public API)

File: keras_cv/models/segmentation/__internal__/segmentation_head.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                tf.keras.layers.Conv2D(
+                    name=conv_name,
+                    filters=self.channels,
+                    kernel_size=3,
+                    padding="same",
+                    use_bias=False,
+                )
+            )
+            norm_name = "segmentation_head_norm_{}".format(i)
+            self._bn_layers.append(tf.keras.layers.BatchNormalization(name=norm_name))
```
Comment (tanzhenyu) [inline keras_cv/models/segmentation/__internal__/segmentation_head.py:72]: should BN be after each conv layer?

File: keras_cv/models/segmentation/__internal__/segmentation_head.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            name="segmentation_output",
+            filters=self.classes,
+            kernel_size=1,
+            padding="same",
+        )
+
+    def call(self, inputs):
+        """Forward path for the segmentation head.
+
+        For now, it accepts the output from the decoder only, which is a dict with int
```
Comment (tanzhenyu) [inline keras_cv/models/segmentation/__internal__/segmentation_head.py:84]: specify the shape of decoder output here?

**Review by tanzhenyu — COMMENTED** [review]
Some minor comments.

File: keras_cv/models/segmentation/__internal__/segmentation_head.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/models/segmentation/__internal__/segmentation_head.py:27]: Good point - if this is only to be used in segmentation asking for foreground classes is a nice way to keep it clear what the user is to pass.  Another approach would be to call this output_dim instead of classes

Comment (qlzh727) [reply keras_cv/models/segmentation/__internal__/segmentation_head.py:27]: Done.

File: keras_cv/models/segmentation/__internal__/segmentation_head.py (hunk 2 above)
Comment (qlzh727) [reply keras_cv/models/segmentation/__internal__/segmentation_head.py:72]: There is one BN per conv2D the stacking, the final classification conv doesn't have a BN.

File: keras_cv/models/segmentation/__internal__/segmentation_head.py (hunk 3 above)
Comment (qlzh727) [reply keras_cv/models/segmentation/__internal__/segmentation_head.py:84]: Oh, it is included in the docstring above (as the code example.)

File: keras_cv/models/segmentation/__internal__/segmentation_head.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    The head will take the output from decoder (eg FPN or ASPP), and produce production
+    (pixel level classifications) as the output for the model.
+
+    Args:
+        classes: int, the number of output classes for the prediction. This should
+            include all the classes (eg background) for the model to predict.
+        convs: int, the number of conv2D layers that are stacked before the final
+            classification layer. Default to 2.
+        channels: int, the number of filter/channels for the the conv2D layers. Default
```
Comment (LukeWood) [inline keras_cv/models/segmentation/__internal__/segmentation_head.py:31]: Nit: Let’s rename to filters because that’s how it is named in conv2d

File: keras_cv/models/segmentation/__internal__/segmentation_head.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+import tensorflow as tf
+from tensorflow.keras import layers
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class SegmentationHead(layers.Layer):
+    """Prediction head for the segmentation model
+
+    The head will take the output from decoder (eg FPN or ASPP), and produce production
```
Comment (ianstenbit) [inline keras_cv/models/segmentation/__internal__/segmentation_head.py:23]: nit: should this say "produce a segmentation mask"?

**Review by ianstenbit — APPROVED** [review]

File: keras_cv/models/segmentation/__internal__/segmentation_head.py (hunk 4 above)
Comment (qlzh727) [reply keras_cv/models/segmentation/__internal__/segmentation_head.py]: Done.

File: keras_cv/models/segmentation/__internal__/segmentation_head.py (hunk 5 above)
Comment (qlzh727) [reply keras_cv/models/segmentation/__internal__/segmentation_head.py]: Done.

**Review by tanzhenyu — APPROVED** [review]
LGTM.
Thanks for the PR!

**Review by LukeWood — APPROVED** [review]
Looks good, thanks Scott

i wonder if we should export these at some point.  Users customizing their models with regularization and such may want to use these components.  Another approach is to just write tutorials on it…

**Comment by qlzh727** [comment]
> Looks good, thanks Scott
> 
> i wonder if we should export these at some point. Users customizing their models with regularization and such may want to use these components. Another approach is to just write tutorials on it…

yes, once we verify this with more configs/use case, this will become public.
